File: java/lib/src/test/java/org/sammancoaching/library/IsbnTest.java
package org.sammancoaching.library;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class IsbnTest {

    @Test
    void testIsbnDigits() {
        Isbn isbn = Isbn.of("978-0201485677");

        assertEquals("9780201485677", isbn.digits());
    }

    @Test
    void testIsbnFormat() {
        Isbn isbn = Isbn.of("978 0201485677");

        assertEquals("9780201485677", isbn.digits());
    }

    @Test
    void testIsbnEquality() {
        assertEquals(Isbn.of("978-0201485677"), Isbn.of("9780201485677"));
    }

    @Test
    void testIsbnInequality() {
        assertNotEquals(Isbn.of("978-0201485677"), Isbn.of("978-0134757599"));
    }

    @Test
    void testInvalidInput() {
        Executable parseIsbnWithInvalidChecksum = () -> Isbn.of("978-0201485670");

        assertThrows(IllegalArgumentException.class, parseIsbnWithInvalidChecksum);
    }

    @Test
    void testInvalidInput2() {
        Executable parseNonNumericIsbn = () -> Isbn.of("abc");

        assertThrows(IllegalArgumentException.class, parseNonNumericIsbn);
    }

    @Test
    void badInputIsBadIguess() {
        Executable parseShortIsbn = () -> Isbn.of("978-020148567");

        assertThrows(IllegalArgumentException.class, parseShortIsbn);
    }
}

File: csharp/Library.Tests/IsbnTest.cs
using NUnit.Framework;
using SammanCoaching.Library;

namespace Library.Tests;

public class IsbnTest
{
    [Test]
    public void testIsbnDigits()
    {
        var value = "978-0201485677";

        var isbn = Isbn.Of(value);

        Assert.That(isbn.Digits, Is.EqualTo("9780201485677"));
    }

    [Test]
    public void testIsbnFormat()
    {
        var value = "978 0201485677";

        var isbn = Isbn.Of(value);

        Assert.That(isbn.Digits, Is.EqualTo("9780201485677"));
    }

    [Test]
    public void testIsbnEquality()
    {
        var firstValue = "978-0201485677";
        var secondValue = "9780201485677";

        var firstIsbn = Isbn.Of(firstValue);
        var secondIsbn = Isbn.Of(secondValue);

        Assert.That(firstIsbn, Is.EqualTo(secondIsbn));
    }

    [Test]
    public void testIsbnInequality()
    {
        var firstValue = "978-0201485677";
        var secondValue = "978-0134757599";

        var firstIsbn = Isbn.Of(firstValue);
        var secondIsbn = Isbn.Of(secondValue);

        Assert.That(firstIsbn, Is.Not.EqualTo(secondIsbn));
    }

    [Test]
    public void testInvalidInput()
    {
        var value = "978-0201485670";

        Action act = () => Isbn.Of(value);

        Assert.Throws<ArgumentException>(act);
    }

    [Test]
    public void testInvalidInput2()
    {
        var value = "abc";

        Action act = () => Isbn.Of(value);

        Assert.Throws<ArgumentException>(act);
    }

    [Test]
    public void badInputIsBadIguess()
    {
        var value = "978-020148567";

        Action act = () => Isbn.Of(value);

        Assert.Throws<ArgumentException>(act);
    }
}
